Restructure the command line interfce implementation

dnicolodi opened the pull request:
This implements #52, and fixes #3.

dnicolodi commented:
This is to be considered work in progress. Working on implementing the same treatment for the `file` and `extract` commands I tweaked the implementation up a bit. The general structure remains the same though.

dnicolodi commented:
This has grown into a major reorganization of the code. I'm quite happy of the result. The implemented functionality should be the same but I find the code much easier now. There are many commits but the important ones are 3d47e97, 
b2bc8cf, 5816867.

blais commented:
LGTM except for one thing: I'd really much rather we keep the tests right next to their modules and not to a separate directory. All the rest of the Beancount code is that way and it breaks the convention.

dnicolodi commented:
I moved the test to the separate directory for two reasons: compatibility with `unittest` requires the tests module names to be prefixed with `test_` instead than to be suffixed with `_test` and because of the doctests with I don't know how to make fit in the naming scheme you propose.

I started looking into the possibility of running the tests with `unittest` because it turns out that the `pytest` runner is hiding several resource leaks. The tests run with the unittest test runner reported the leaks and failed. At a very superficial look, the leaks were only in the test code, but I am not sure yet, and because of the recent fixes you committed to fix some in the CSV parser I think it would have been nice to catch them early.

blais commented:
These are good reasons but at the same time, it'll be annoying to have different ways to do things. It's nicer to do it all the same across all projects IMO.
I'm rather used to pytest as a default test runner, it's pretty slick, but I'm a dinosaur.
Is it out of fashion?

dnicolodi commented:
I don't think pytest is out of fashion but there are more things I find annoying about it that interesting features. I think that in Python 3 unittest gained the few features that could have made using pytest interesting for me. All projects I work on have tests in a separate directory. pytest discovers tests named in the unittest format just fine. What about moving the test alongside the code but named with a `test_` prefix instead than a `_test.py` suffix?

dnicolodi commented:
I moved the tests next to the modules they are testing, except for the doctests, these I kept in the `bangulp/tests/` directory. I renamed the `file` command to `archive` as discussed.

blais commented on beangulp/extract.py:
```diff
@@ -1,184 +1,114 @@
+
+
+def print_extracted_entries(extracted, output):
+    """Print extracgted entries.
```
typo

blais commented:
> I don't think pytest is out of fashion but there are more things I find annoying about it that interesting features. I think that in Python 3 unittest gained the few features that could have made using pytest interesting for me. All projects I work on have tests in a separate directory. pytest discovers tests named in the unittest format just fine. What about moving the test alongside the code but named with a `test_` prefix instead than a `_test.py` suffix?

Thank you Daniele!

If we use a suffix when you list the files in alphabetical order the implementation and the test sit right next to each other and it makes it obvious what doesn't have a test. It's not that super important, just another convention. The rest of the Beancount universe is named that way... I don't care all that much, though I do like things to be more like each other than different from each other unless there's some reason for it.

If you really care a lot for it, name it the way you like.

dnicolodi commented:
I proposed the different naming because `test_*.py` is what `unittest` expects and what most projects use. I was not aware that there is a command line switch to modify the file name glob pattern used by `unittest`. It may have been added relatively recently. With the tests next to the sources I prefer the current naming.

To run the tests with unittest, `python -m unittest discover -p '*_test.py'` is not much longer than `python -m unittest` which would suffice with the conventional unittest test naming.
